---
title: "Cho_Fig4C_BulkRNA.RMD"
author: "Brian Muegge"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# The RMD file is in subdirectory scripts/. Change root directory to the main directory
knitr::opts_knit$set(root.dir = "..")
```

# Overview

To explore the physiologic role of motile primary cilia in human islets and potential link to metabolic disease, we examined ciliary gene expression in islet transcriptome studies from both healthy donors and individuals with type 2 diabetes.

This code is the complete workflow to generate Figure 4C of the paper, beginning with quasimapped reads from the publicly available Arda, et al, 2016 study.

## Preparation

Raw fastq files from a published RNA-seq study of isolated pancreatic cells were downloaded from the European Nucleotide Archive (Arda et al 2016; PMC4864151 ; 10.1016/j.cmet.2016.04.002 ; Submission ID SRA167055)

Reads were quasi-mapped to the human genome (GRCh38, Ensembl release 106) using Salmon version 1.3.0 (Patro, et al 2017; https://doi.org/10.1038/nmeth.4197).

* The genome for Homo sapiens gc38 was downloaded from Ensembl in April 2022 (http://ftp.ensembl.org/pub/release-106/fasta/homo_sapiens/).

* The salmon index was generated with the following command:
*salmon index -t Homo_sapiens.GRCh38.cdna.all.fa.gz -i <path_to_make_index> -k 31*

* Paired end reads were quasimapped with salmon:
*salmon quant -i $INDEX_PATH -l A -1 <read1> -2 <read2> -o <salmon_out> --useVBOpt --gcBias --validateMappings --quiet*

## Block 1: Establish File Structure, Load Libraries, set Variables

### Paths and Files block

```{r parameters}
myWorkingDirectory = getwd()
figureOutputDirectory = file.path(myWorkingDirectory,"figures")
analysisOutputDirectory = file.path(myWorkingDirectory,"analysis")
metadataDirectory = file.path(myWorkingDirectory,"metadata")
dataDirectory = file.path(myWorkingDirectory,"data")
deOutputDirectory = file.path(myWorkingDirectory,"diffexp")
scriptsDirectory = file.path(myWorkingDirectory,"scripts")

# Metadata
sampleMetadataFile <- file.path(metadataDirectory,"Arda_SampleMetadata.txt")
```

### R Libraries and functions

```{r source}
library(AnnotationHub)
library(tximport)

library(limma)
library(edgeR)
library(biomaRt)

library(knitr)

library(pheatmap)
library(RColorBrewer)
library(ggsci)

library(ggplot2)
```

### Values for Differential Expression Testing

```{r}
maxFDR <- 0.05
minLFC <- 1
```

### Color Palettes and Graphics
```{r graphics}
paletteCells.four <- pal_nejm()(4)

# http://www.cookbook-r.com/Graphs/Colors_(ggplot2)/
palette_cb_gray <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

```



## Block 2: Load Salmon results and generate gene annotation

### Load the metadata file

```{r}
sampleData <- read.delim(sampleMetadataFile,header = TRUE,row.names = 1,stringsAsFactors = FALSE)
```

### Use tximport to load the Salmon pseudoaligned counts

Here I get version 105 of human genome

https://support.bioconductor.org/p/109092/

```{r}
library(AnnotationHub)
ah <- AnnotationHub()
query(ah,pattern=c("EnsDb.Hsapiens"))
edb <- ah[["AH98047"]]

## Now get the transcripts table as a data frame

txs <- transcripts(edb, return.type = "DataFrame")
dim(txs)
colnames(txs)
```

tx2gene argument requires a two-column data.frame linking transcript id (column 1) to gene id (column 2). the column names are not relevant, but this column order must be used.

```{r}
txs.small <- data.frame(txs$tx_id_version, txs$gene_id)
```

Confirm that all the files exist

```{r}
file.exists(sampleData$files)
```

Finally, use tximport to process the quant.sf files

https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html

Because my usual workflow is Limma-voom, I will scale the counts at creation using lengthScaledTPM as recommended by the package authors:
https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html#limma-voom

```{r}
txi <- tximport(sampleData$files, type = "salmon", tx2gene = txs.small, countsFromAbundance = "lengthScaledTPM")
```

### Generate Gene Annotation

The salmon data was generated with Ensembl id. Use standardized system to expand annotation

```{r}
# Building on code shared by Kevin Blighe, https://www.biostars.org/p/388949/#388960
  
mart <- useMart('ENSEMBL_MART_ENSEMBL')
mart <- useDataset('hsapiens_gene_ensembl', mart)
  
# searchAttributes(mart,pattern = "symbol")
annotLookup <- getBM(
    mart = mart,
    attributes = c(
      'wikigene_description',
      'ensembl_gene_id',
      'entrezgene_id',
      'gene_biotype',
      'external_gene_name'))
  
# Use the ensembl keys (row names of the abundance table) to build the annotation
annotLookup <- annotLookup[which(annotLookup$ensembl_gene_id %in% row.names(txi$abundance)),]
annotLookup <- annotLookup[match(row.names(txi$abundance), annotLookup$ensembl_gene_id),]
# check that annots are aligned
all.equal(row.names(txi$abundance), annotLookup$ensembl_gene_id) 
  
geneData <- data.frame(matrix(nrow =length(row.names(txi$abundance)),ncol=5))
row.names(geneData) <- row.names(txi$abundance)
colnames(geneData) <- c("GENENAME","ENTREZID","BIOTYPE","SYMBOL","ENSEMBLID")
geneData$GENENAME <- annotLookup$wikigene_description
geneData$ENTREZID <- annotLookup$entrezgene_id
geneData$BIOTYPE <- annotLookup$gene_biotype
geneData$SYMBOL <- annotLookup$external_gene_name
geneData$ENSEMBLID <- row.names(txi$abundance)
  
# replace blanks with NA, and <NA> with "NA"
geneData[geneData==""] <- "NA"
geneData[is.na(geneData)] <- "NA"
```


## Block 3: Create Limma object and perform basic QC

### Create a limma object

```{r createLimma}

# build the limma object. Use remove.zeros= FALSE to keep congruence with genes file
y <- DGEList(counts = txi$counts,
             samples = sampleData
             ,genes = geneData,
             remove.zeros = FALSE,
             group = sampleData$Group)

cat("Before QC and filtering, there are",ncol(y),"samples and",nrow(y),"genes in the dataset.\n\n")
```

## Block 4: Filter out very lowly expressed genes

Minimal filter requiring log2(counts) / million >= 0 in at least 4 samples (minimum group size =4)

```{r filterLowExpressed}

myMinCPM <- 0
myMinSample <- 4

keep.exprs <- rowSums(cpm(y,log=TRUE) >=myMinCPM) >=myMinSample
sum(keep.exprs)
  
x <- y[keep.exprs,keep.lib.sizes=FALSE]
x <- calcNormFactors(x)
  
cat("After filtering for genes that have at least", myMinCPM,
      "cpm in at least", myMinSample, "samples, there are",   
      dim(x)[1], "genes retained in the DGElist.\n\n")
```

## Block 5: Sample Heatmap Clustering

```{r}
sampleDists <- dist(t(cpm(x,log=TRUE)),method="euclidean")

# Note that to work with annotation, the values must be in a data.frame with row.names
sampleDistMatrix <- as.matrix(sampleDists)
sampleDistMatrix <- as.data.frame(sampleDistMatrix)

# Color Ramp. Dark blue for most similar, white for most dissimilar
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)

# Annotation DF with metadata on the samples
#  First, pull only columns from the samples metadata that are relevant to plot
annotationDF <- x$samples[,colnames(x$samples) %in% c("Group","age_group")]
rownames(annotationDF) <- rownames(sampleDistMatrix)

# Reorder so Group(Region) on the bottom
annotationDF <- annotationDF[,c(2,1)]
rownames(annotationDF) <- rownames(sampleDistMatrix)
colnames(annotationDF) <- c("Cells","Age_Group")

anno_colors <- list(
  Cells = paletteCells.four,
  Age_Group = palette_cb_gray[c(1,5)] # gray, light blue
)
names(anno_colors$Cells) <- sort(unique(x$samples$Group))
names(anno_colors$Age_Group) <- sort(unique(x$samples$age_group))

p <- pheatmap(sampleDistMatrix,
         col = colors,
         border_color="gray10",
         show_rownames=FALSE,
         show_colnames = FALSE,
         annotation_colors = anno_colors,
         annotation_col = annotationDF,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists)
p
```

## Block 6: Differential Expression Testing

```{r}
group <- factor(x$samples$Group)

design <- model.matrix(~0+group)
colnames(design) <- levels(group)

# make a contrast matrix
cont.matrix <- makeContrasts(
  Beta_vs_Alpha = "beta-alpha",
  Beta_vs_Acinar = "beta-acinar",
  Beta_vs_Duct = "beta-duct",
  Alpha_vs_Acinar = "alpha-acinar",
  Alpha_vs_Duct = "alpha-duct",
  Duct_vs_Acinar = "duct-acinar",
  levels = design)

# Limma testing
v <- voom(x, design, plot=FALSE)
fit <- lmFit(v, design)
fit2  <- contrasts.fit(fit, cont.matrix)
efit.all  <- eBayes(fit2)

# Summarize DE testing results
dt <- decideTests(efit.all,p.value=maxFDR,adjust.method = "BH",lfc = minLFC)
cat("Considering RNA seq dataset with BH adjusted max p-value",maxFDR,"and minimum log2 fold change",minLFC,"\n")

kable(summary(dt))
```


## Block 7: Implement FGSEA with Cilia Lists,

### Convert the elist into objects for use in fgsea

```{r}
library(fgsea)
library(dplyr)
library(tidyverse)
library(grid)

# Get the beta cell vs acinar comparison
myAnno <- x$genes[,c("SYMBOL","GENENAME","ENTREZID","BIOTYPE")]

myTopTable <- topTable(efit.all,coef = 2, number = nrow(dt),sort.by = "none",genelist = myAnno)

# Tidy the row.names
myTopTable.tibble <- rownames_to_column(myTopTable)

# Get extracted tibbles for logFC and t
tt.logFC <- myTopTable %>% 
    dplyr::select(SYMBOL, logFC) %>% 
    na.omit() %>% 
    distinct() %>% 
    group_by(SYMBOL)

tt.tTest <- myTopTable %>% 
    dplyr::select(SYMBOL, t) %>% 
    na.omit() %>% 
    distinct() %>% 
    group_by(SYMBOL)

# make ranks
ranks.logFC <- deframe(tt.logFC)
ranks.tTest <- deframe(tt.tTest)
```

### Make a gene set of cilia genes

CiliaCarta includes different levels of evidence - "Gold Standard", "Gene Ontology", and "Predicted". 

They also generate a "score". A log likelihood score of 1 means the gene is twice as like to be ciliary as not ciliary.

We use genes with score greater than 1 as our gene set.

```{r}
ciliaCarta <- read.delim(file.path(dataDirectory,"CiliaCarta.csv"),header=TRUE,row.names = NULL,sep=",",stringsAsFactors = FALSE)

ciliaCarta$ScoreGT1[ciliaCarta$CiliaCarta.Score>=1] <- TRUE
 
pathways.cilia <- list("HighScore"= ciliaCarta$Associated.Gene.Name[ciliaCarta$ScoreGT1==TRUE])
```

```{r}
# run the analysis
fgseaRes <- fgsea(pathways=pathways.cilia, stats=ranks.tTest)

#Tidy the results
fgseaResTidy <- fgseaRes %>%
    as_tibble() %>%
    arrange(desc(NES))

head(fgseaResTidy)

```


Make the High Score Plot

```{r}
# Use FGSEA function to generate the GSEA plot, then
#  customize appearance
temp.gg <- plotEnrichment(pathways.cilia[["HighScore"]],
                          ranks.logFC) + 
                          labs(title="High Score Gene Set")

# pull stats (Pval, padj, ES, NES). 
pathwayGSEAstats <- data.frame( 
  fgseaResTidy[ fgseaResTidy$pathway == "HighScore", 2:6 ])

# Make a text box
textBox <- paste0( "NES: ", 
                   sprintf("%.2f",pathwayGSEAstats$NES), 
                   "\n","FDR < 0.001\n")

grob <- textGrob(textBox, x=0.95,  y=0.8, hjust=1,
                 gp=gpar(col="black", 
                         fontsize=13,fontface="bold"))

# Pull the positions for the vertical line segments. 
# The x coordinates for segment ticks are stored in $data
pathwaySegmentsX <- as.data.frame(temp.gg$layers[[6]]$data)

# Pull data limits
xMin <- 0
xMax <- max(temp.gg$data$x)
yMin <- min(temp.gg$data$y)
yMax <- max(temp.gg$data$y)

# make a dummy data frame for color bar
# segments will be 0.2 below y, color bar will centered on 0.3 below yMin
colorDF <- data.frame(dummy=seq(xMin:xMax),fixedY=rep(yMin-0.3,xMax+1))

g <-ggplot(temp.gg$data, aes(x=x, y=y)) +
  coord_cartesian(xlim = c(0,xMax), ylim= 
                  c(yMin-0.4,yMax*1.05), expand = FALSE) +
  geom_point(color="red", size=0.1) +
  geom_line(color="red") +
  geom_hline(yintercept=0, colour="black", 
             linetype="dashed") +
  labs(x="Rank in Ordered Gene List", 
       y="Enrichment Score (ES)") +
  geom_tile(data=colorDF, 
            aes(x=dummy,y=fixedY, 
                height=0.2,fill = dummy)) +
  # hex code from brewer.pal(9,"Reds") 
  #   and brewer.pal(9,"Blues")
  scale_fill_gradientn( 
    values=c(0,0.15,0.3,0.5,0.7,0.85,1),                  
    colours=c("red","#FCBBA1","#FEE0D2", 
              "white","#DEEBF7","#C6DBEF","dodgerblue2")) +
  geom_segment(data=pathwaySegmentsX,
               mapping=aes(x=x, y=-yMin-0.2,
                           xend=x, yend=yMin),
               size=0.3) +
  geom_rect(aes(xmin = 0, xmax = xMax, 
                ymin = yMin-0.2, ymax = yMin),
            fill = "transparent", color = "black", 
            size = 0.5) +
  geom_rect(aes(xmin = 0, xmax = xMax, 
                ymin = yMin-0.4, ymax = yMin-0.2),
            fill = "transparent", color = "black", 
            size = 0.5) +
  theme_bw() +
  theme(legend.position="none", 
        axis.text = element_text(size=12,face="bold"),
        axis.title = element_text(size=14,face="bold"))

g + annotation_custom(grob)

pdf(file=file.path(figureOutputDirectory,"Fig4C_HighScore_GSEA_final.pdf"),width = 3.5, height = 3)
g + annotation_custom(grob)
dev.off()
```

```{r}
sessionInfo()
```